File: backtracking/_40_Combination_Sum_II.java
package backtracking;

import java.util.*;

public class _40_Combination_Sum_II {
    public static void main(String[] args) {

    }


    class Solution {
        List<List<Integer>> res = new ArrayList<>();

        public List<List<Integer>> combinationSum2(int[] candidates, int target) {
            Arrays.sort(candidates); // sort for duplicate handling
            dfs(candidates, 0, target, new ArrayList<>());
            return res;
        }

        private void dfs(int[] nums, int index, int remain, List<Integer> path) {
            if (remain == 0) {
                res.add(new ArrayList<>(path));
                return;
            }
            if (index >= nums.length || remain < 0) return;

            // Option 1: take this number
            path.add(nums[index]);
            dfs(nums, index + 1, remain - nums[index], path);
            path.remove(path.size() - 1);

            // Option 2: skip all duplicates of this number
            int next = index + 1;
            while (next < nums.length && nums[next] == nums[index]) {
                next++;
            }
            dfs(nums, next, remain, path);
        }
    }

}

File: backtracking/_40_Combination_Sum_II.py
class Solution:

    def combinationSum2(self, candidates, target: int):
        self.my_set = set()
        candidates.sort()
        self.helper(0, target, [], candidates)
        return list(self.my_set)

    def helper(self, index, target, current, arr):
        if target == 0:
            self.my_set.add(tuple(current))
            return

        if target < 0 or index >= len(arr):
            return

        current.append(arr[index])
        self.helper(index + 1, target - arr[index], current, arr)
        current.pop()
        self.helper(index + 1, target, current, arr)

a = Solution()
ab=a.combinationSum2([2,5,2,1,2],5)
print(ab)